subsys/net/ip/net_tls.c

```diff
@@ -0,0 +1,735 @@
+				context->tls->rx_pkt->frags->data;
+		} else {
+			return MBEDTLS_ERR_SSL_WANT_READ;
+		}
+	}
+
+	len_left = net_pkt_get_len(context->tls->rx_pkt) -
+		   context->tls->rx_offset;
+	ret = net_frag_linearize(buf, len, context->tls->rx_pkt,
+				 context->tls->rx_offset,
+				 MIN(len, len_left));
+	if (ret > 0) {
+		context->tls->rx_offset += ret;
+	}
+
+	if (context->tls->rx_offset >=
+	    net_pkt_get_len(context->tls->rx_pkt)) {
+		if (net_context_get_type(context) == SOCK_STREAM) {
+			net_context_update_recv_wnd(
+				context,
+				net_pkt_appdatalen(context->tls->rx_pkt));
+		}
+		net_pkt_unref(context->tls->rx_pkt);
+		context->tls->rx_pkt = NULL;
+		context->tls->rx_offset = 0;
+	}
+
+	return ret;
+}
+
+static void tls_recv_cb(struct net_context *context, struct net_pkt *pkt,
+			int status, void *user_data)
+{
+	struct net_pkt *decrypted_pkt;
+	int len, ret = 0;
+
+	if (!pkt) {
+		/* Forward EOF */
+		if (context->tls->tls_cb) {
+			context->tls->tls_cb(context, pkt, status, user_data);
+		}
+		return;
+	}
+
+	k_fifo_put(&context->tls->rx_fifo, pkt);
+
+	if (net_context_get_type(context) == SOCK_STREAM) {
+		net_context_update_recv_wnd(context, -net_pkt_appdatalen(pkt));
+	}
+
+	/* Process it as application data only after handshake is over,
+	 * otherwise packet will be consumed by handshake.
+	 */
+	if (context->tls->ssl.state == MBEDTLS_SSL_HANDSHAKE_OVER) {
+		len = net_pkt_appdatalen(pkt);
+
+		decrypted_pkt = net_pkt_clone(pkt, 0, NET_PKT_CLONE_HDR);
+		if (!decrypted_pkt) {
+			return;
+		}
+
+		/* Copy IP and transport header */
+		u16_t header_len = net_pkt_appdata(pkt) - pkt->frags->data;
+
+		ret = net_frag_linear_copy(decrypted_pkt->frags, pkt->frags,
+					   0, header_len);
+		if (ret < 0) {
+			goto err;
+		}
+
+		/* Set appdata and appdatalen */
+		net_pkt_set_appdata(decrypted_pkt,
+				    decrypted_pkt->frags->data + header_len);
+		net_pkt_set_appdatalen(decrypted_pkt, 0);
+
+		while (1) {
+			ret = mbedtls_ssl_read(&context->tls->ssl,
+					       context->tls->rx_ssl_buf,
+					       sizeof(context->tls->rx_ssl_buf));
+			if (ret == 0 || ret == MBEDTLS_ERR_SSL_WANT_READ ||
+				ret == MBEDTLS_ERR_SSL_WANT_WRITE) {
+				ret = 0;
+
+				break;
+			}
+
+			if (ret > 0) {
+				ret = net_pkt_append(decrypted_pkt, ret,
+						     context->tls->rx_ssl_buf,
+						     TIMEOUT_TLS_RX_MS);
+				/* Update appdatalen */
+				net_pkt_set_appdatalen(decrypted_pkt,
+						       net_pkt_appdatalen(decrypted_pkt) + ret);
+				continue;
+			}
+
+			ret = -EBADF;
+			break;
+		}
+
+		if (ret >= 0 && context->tls->tls_cb) {
+			context->tls->tls_cb(context, decrypted_pkt,
+					     status, user_data);
+		} else {
+			goto err;
+		}
+	}
+
+	return;
+
+err:
+	net_pkt_unref(decrypted_pkt);
+}
+
+static int tls_add_ca_certificate(struct net_context *context,
+				  struct net_tls_credential *ca_cert)
+{
+#if defined(MBEDTLS_X509_CRT_PARSE_C)
+	int err = mbedtls_x509_crt_parse(&context->tls->ca_chain,
```

Comment:
As a note, docstring of mbedtls_x509_crt_parse():

~~~
 * \brief          Parse one or more certificates and add them
 *                 to the chained list. Parses permissively. If some
 *                 certificates can be parsed, the result is the number
 *                 of failed certificates it encountered. If none complete
 *                 correctly, the first error is returned.
~~~

I.e. it can return error, and still land some certs in the internal state. @d3zd3z .

(Yes, *this* can be fixed later, now just TODO needs to be added.)

Reply:
Yes, but in this case we parse one certificate at a time, (`tls_add_ca_certificate` takes single `ca_cert`, not a chain), so I'm not sure if there is anything to fix here.